tikv/raftstore/util.go

```diff
@@ -360,3 +362,89 @@ func isVoteMessage(msg *eraftpb.Message) bool {
 	tp := msg.GetMsgType()
 	return tp == eraftpb.MessageType_MsgRequestVote || tp == eraftpb.MessageType_MsgRequestPreVote
 }
+
+func regionIDToBytes(id uint64) []byte {
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, id)
+	return b
+}
+
+func regionIDFromBytes(b []byte) uint64 {
+	return binary.LittleEndian.Uint64(b)
+}
+
+func checkRegionEpoch(req *raft_cmdpb.RaftCmdRequest, region *metapb.Region, includeRegion bool) error {
+	var checkVer, checkConfVer bool
+	if req.AdminRequest == nil {
+		// for get/set/delete, we don't care conf_version.
+		checkVer = true
+	} else {
+		switch req.AdminRequest.CmdType {
+		case raft_cmdpb.AdminCmdType_CompactLog, raft_cmdpb.AdminCmdType_InvalidAdmin,
+			raft_cmdpb.AdminCmdType_ComputeHash, raft_cmdpb.AdminCmdType_VerifyHash:
+		case raft_cmdpb.AdminCmdType_ChangePeer:
+			checkConfVer = true
+		case raft_cmdpb.AdminCmdType_Split, raft_cmdpb.AdminCmdType_BatchSplit,
+			raft_cmdpb.AdminCmdType_PrepareMerge, raft_cmdpb.AdminCmdType_CommitMerge,
+			raft_cmdpb.AdminCmdType_RollbackMerge, raft_cmdpb.AdminCmdType_TransferLeader:
+			checkVer = true
+			checkConfVer = true
+		}
+	}
+	if !checkVer && !checkConfVer {
+		return nil
+	}
+	if req.Header.RegionEpoch == nil {
+		return errors.Errorf("missing epoch")
+	}
+
+	fromEpoch := req.Header.RegionEpoch
+	currentEpoch := region.RegionEpoch
+
+	// We must check epochs strictly to avoid key not in region error.
+	//
+	// A 3 nodes TiKV cluster with merge enabled, after commit merge, TiKV A
+	// tells TiDB with a epoch not match error contains the latest target Region
+	// info, TiDB updates its region cache and sends requests to TiKV B,
+	// and TiKV B has not applied commit merge yet, since the region epoch in
+	// request is higher than TiKV B, the request must be denied due to epoch
+	// not match, so it does not read on a stale snapshot, thus avoid the
+	// KeyNotInRegion error.
+	if (checkConfVer && fromEpoch.ConfVer != currentEpoch.ConfVer) ||
+		(checkVer && fromEpoch.Version != currentEpoch.Version) {
+		err := &ErrEpochNotMatch{}
+		if includeRegion {
+			err.Regions = []*metapb.Region{region}
```

Comment:
I think you should clone region here?

Reply:
Right, I `to_own` calls `clone`, I didn't noticed that before.